File: Skin Editor/SkinEditor/SkinEditor/Host.vb
Imports System.Drawing.Imaging
Imports SkinEditor.XPRDecoder
Imports SkinEditor.Interfaces

Public Class Host
    Implements IHost

    Public Event ErrorEvent(ByVal ErrorLevel As Integer, ByVal Message As String, ByVal File As String, ByVal Line As Integer, ByVal Pos As Integer)

    Private objSettings As New Settings

    Public Function AppPath() As String
        Return System.AppDomain.CurrentDomain.BaseDirectory()
    End Function

#Region "Misc Functions"

    Public Sub ShowMessage(ByVal Message As String) Implements IHost.ShowMessage
        MsgBox(Message)
    End Sub

    Public Sub ShowForm(ByVal Form As Form, ByVal DockLocation As String) Implements IHost.ShowForm
        Form.Show()
    End Sub

    Public Function ShowDialog(ByVal Form As Form) As DialogResult Implements IHost.ShowDialog
        Return Form.ShowDialog()
    End Function

    Public Sub ErrorOccured(ByVal ErrorLevel As Integer, ByVal Message As String, ByVal File As String, ByVal Line As Integer, ByVal Pos As Integer) Implements IHost.ErrorOccured
        RaiseEvent ErrorEvent(ErrorLevel, Message, File, Line, Pos)
    End Sub

    Public Function ExecuteCommand(ByVal Command As String, ByVal Parameters() As Object) As Object() Implements IHost.ExecuteCommand
        Return Nothing
    End Function

#End Region

#Region "XPR Functions"

    Public Function GetXPRImageCount(ByVal FileName As String) As Integer Implements IHost.GetXPRImageCount
        Dim objDecoder As New Decoder
        objDecoder.OpenXPR(FileName)
        GetXPRImageCount = objDecoder.FileCount
        objDecoder.CloseXPR()
    End Function

    Public Function GetXPRImageName(ByVal FileName As String, ByVal ImageIndex As Integer) As String Implements IHost.GetXPRImageName
        Dim objDecoder As New Decoder
        objDecoder.OpenXPR(FileName)
        GetXPRImageName = objDecoder.FileName(ImageIndex)
        objDecoder.CloseXPR()
    End Function

    Public Function GetXPRImageLoops(ByVal FileName As String, ByVal ImageIndex As Integer) As Integer Implements IHost.GetXPRImageLoops
        Dim objDecoder As New Decoder
        objDecoder.OpenXPR(FileName)
        GetXPRImageLoops = objDecoder.GetImage(ImageIndex).Loops
        objDecoder.CloseXPR()
    End Function

    Public Function GetXPRImageFrameCount(ByVal FileName As String, ByVal ImageIndex As Integer) As Integer Implements IHost.GetXPRImageFrameCount
        Dim objDecoder As New Decoder
        objDecoder.OpenXPR(FileName)
        GetXPRImageFrameCount = objDecoder.GetImage(ImageIndex).Frames.Count
        objDecoder.CloseXPR()
    End Function

    Public Function GetXPRImageFrameDelay(ByVal FileName As String, ByVal ImageIndex As Integer, ByVal FrameIndex As Integer) As Integer Implements IHost.GetXPRImageFrameDelay
        Dim objDecoder As New Decoder
        objDecoder.OpenXPR(FileName)
        GetXPRImageFrameDelay = objDecoder.GetImage(ImageIndex).Frames(FrameIndex).Delay
        objDecoder.CloseXPR()
    End Function

    Public Function GetXPRImageFrame(ByVal FileName As String, ByVal ImageIndex As Integer, ByVal FrameIndex As Integer) As Image Implements IHost.GetXPRImageFrame
        Dim objDecoder As New Decoder
        objDecoder.OpenXPR(FileName)
        GetXPRImageFrame = objDecoder.GetImage(ImageIndex).Frames(FrameIndex).Frame
        objDecoder.CloseXPR()
    End Function

    Public Sub SaveXPRImage(ByVal FileName As String, ByVal ImageIndex As Integer, ByVal DestFileName As String) Implements IHost.SaveXPRImage
        Dim objDecoder As New Decoder
        objDecoder.OpenXPR(FileName)
        objDecoder.SaveImage(ImageIndex, DestFileName)
        objDecoder.CloseXPR()
    End Sub

    Public Sub SaveXPRImageFrame(ByVal FileName As String, ByVal ImageIndex As Integer, ByVal FrameIndex As Integer, ByVal DestFileName As String) Implements IHost.SaveXPRImageFrame
        Dim objDecoder As New Decoder
        objDecoder.OpenXPR(FileName)
        objDecoder.GetImage(ImageIndex).Frames(FrameIndex).Frame.Save(DestFileName)
        objDecoder.CloseXPR()
    End Sub

#End Region

#Region "GIF Functions"

    Public Function GetGIFImageLoops(ByVal FileName As String) As Integer Implements IHost.GetGIFImageLoops

        Dim objImage As Drawing.Image = Drawing.Image.FromFile(FileName)
        Dim objDimension As New FrameDimension(objImage.FrameDimensionsList(0))

        GetGIFImageLoops = 0

        Dim objPropertyItem As PropertyItem
        For Each objPropertyItem In objImage.PropertyItems
            Select Case objPropertyItem.Id
                Case 20737
                    GetGIFImageLoops = BitConverter.ToInt16(objImage.GetPropertyItem(objPropertyItem.Id).Value, 0)
            End Select
        Next objPropertyItem

        objImage.Dispose()

    End Function

    Public Function GetGIFImageFrameCount(ByVal FileName As String) As Integer Implements IHost.GetGIFImageFrameCount

        Dim objImage As Drawing.Image = Drawing.Image.FromFile(FileName)
        Dim objDimension As New FrameDimension(objImage.FrameDimensionsList(0))
        GetGIFImageFrameCount = objImage.GetFrameCount(objDimension)
        objImage.Dispose()

    End Function

    Public Function GetGIFImageFrameDelay(ByVal FileName As String, ByVal FrameIndex As Integer) As Integer Implements IHost.GetGIFImageFrameDelay

        Dim bytDelays() As Byte = Nothing
        Dim objImage As Drawing.Image = Drawing.Image.FromFile(FileName)
        Dim objDimension As New FrameDimension(objImage.FrameDimensionsList(0))

        Dim objPropertyItem As PropertyItem
        For Each objPropertyItem In objImage.PropertyItems
            Select Case objPropertyItem.Id
                Case 20736
                    bytDelays = objImage.GetPropertyItem(objPropertyItem.Id).Value
            End Select
        Next objPropertyItem

        If bytDelays Is Nothing Then
            GetGIFImageFrameDelay = 0
        Else
            GetGIFImageFrameDelay = BitConverter.ToInt32(bytDelays, FrameIndex * 4) * 10
        End If

        objImage.Dispose()

    End Function

    Public Function GetGIFImageFrame(ByVal FileName As String, ByVal FrameIndex As Integer) As Image Implements IHost.GetGIFImageFrame

        Dim objImage As Drawing.Image = Drawing.Image.FromFile(FileName)
        Dim objDimension As New FrameDimension(objImage.FrameDimensionsList(0))
        Dim objMemoryStream As New System.IO.MemoryStream()

        objImage.SelectActiveFrame(objDimension, FrameIndex)
        objImage.Save(objMemoryStream, System.Drawing.Imaging.ImageFormat.Png)
        GetGIFImageFrame = System.Drawing.Image.FromStream(objMemoryStream)

        objImage.Dispose()

    End Function

#End Region

#Region "XML Functions"

    Public Function XMLClean(ByVal XML As String) As String Implements IHost.XMLClean
        Dim objXMLTools As New XMLTools
        Return objXMLTools.CleanXML(XML)
    End Function

    Public Function XMLValidate(ByVal XML As String) As Boolean Implements IHost.XMLValidate

    End Function

    Public Function XMLMergeIncludes(ByVal IncludeXML As String, ByVal XML As String) As String Implements IHost.XMLMergeIncludes

    End Function

    Public Function XMLMergeReferences(ByVal ReferencesXML As String, ByVal XML As String) As String Implements IHost.XMLMergeReferences

    End Function

#End Region

#Region "Settings Functions"

    Public Sub LoadSettings() Implements IHost.LoadSettings
        objSettings.Load(AppPath() & "Settings.xml")
    End Sub

    Public Sub SaveSettings() Implements IHost.SaveSettings
        objSettings.Save(AppPath() & "Settings.xml")
    End Sub

    Public Function GetSetting(ByVal SettingName As String) As String Implements IHost.GetSetting
        Return objSettings.GetSetting(SettingName)
    End Function

    Public Sub SetSetting(ByVal SettingName As String, ByVal Value As String) Implements IHost.SetSetting
        objSettings.SetSetting(SettingName, Value)
    End Sub

    Public Function GetPluginSetting(ByVal PluginName As String, ByVal SettingName As String) As String Implements IHost.GetPluginSetting
        Return objSettings.GetPluginSetting(PluginName, SettingName)
    End Function

    Public Sub SetPluginSetting(ByVal PluginName As String, ByVal SettingName As String, ByVal Value As String) Implements IHost.SetPluginSetting
        objSettings.SetPluginSetting(PluginName, SettingName, Value)
    End Sub

#End Region

End Class

File: XBMCSkinEditor/Classes/Host.cs
using System.Drawing.Imaging;
using SkinEditor.XPRDecoder;

using XBMCSkinEditor.Properties;
using System.Windows.Forms;
using System.Drawing;

namespace XBMCSkinEditor.Classes
{
    public class Host : SkinEditor.Interfaces.IHost
    {
        
        public event ErrorEventEventHandler ErrorEvent;
        public delegate void ErrorEventEventHandler(int ErrorLevel, string Message, string File, int Line, int Pos);

        private Settings objSettings = new Settings();

        public string AppPath() {
            return System.AppDomain.CurrentDomain.BaseDirectory;
        }
        //Public Function HasSettings(ByVal PluginName As String) Implements IHost.H
        // objSettings.SetPluginSetting(PluginName, SettingName, Value)
        // return objHost.
        //End Function
        #region "Misc Functions"

        public void ShowMessage(string Message)
        {
            System.Windows.Forms.MessageBox.Show(Message);
        }

        public void ShowForm(Form Form, string DockLocation)
        {
            Form.Show();
        }

        public void ShowDockContent(WeifenLuo.WinFormsUI.Docking.DockContent DockCon, WeifenLuo.WinFormsUI.Docking.DockState DDockStyle)
        {
            DockCon.Show(MainWin.mMainWin.DockPanel, DDockStyle);
        }

        public DialogResult ShowDialog(Form Form)
        {
            return Form.ShowDialog();
        }

        public void ErrorOccured(int ErrorLevel, string Message, string File, int Line, int Pos)
        {
            if (ErrorEvent != null)
            {
                ErrorEvent(ErrorLevel, Message, File, Line, Pos);
            }
        }

        public object[] ExecuteCommand(string Command, object[] Parameters)
        {
            return null;
        }

        #endregion

        #region "XPR Functions"

        public int GetXPRImageCount(string FileName)
        {
            int functionReturnValue = 0;
            Decoder objDecoder = new Decoder();
            objDecoder.OpenXPR(FileName);
            functionReturnValue = objDecoder.FileCount;
            objDecoder.CloseXPR();
            return functionReturnValue;
        }

        public string GetXPRImageName(string FileName, int ImageIndex)
        {
            string functionReturnValue = null;
            Decoder objDecoder = new Decoder();
            objDecoder.OpenXPR(FileName);
            functionReturnValue = objDecoder.get_FileName(ImageIndex);
            objDecoder.CloseXPR();
            return functionReturnValue;
        }

        public int GetXPRImageLoops(string FileName, int ImageIndex)
        {
            int functionReturnValue = 0;
            SkinEditor.XPRDecoder.Decoder objDecoder = new SkinEditor.XPRDecoder.Decoder();
            objDecoder.OpenXPR(FileName);
            functionReturnValue = objDecoder.get_GetImage(ImageIndex).Loops;
            objDecoder.CloseXPR();
            return functionReturnValue;
        }

        public int GetXPRImageFrameCount(string FileName, int ImageIndex)
        {
            int functionReturnValue = 0;
            Decoder objDecoder = new Decoder();
            objDecoder.OpenXPR(FileName);
            functionReturnValue = objDecoder.get_GetImage(ImageIndex).Frames.Count;
            objDecoder.CloseXPR();
            return functionReturnValue;
        }

        public int GetXPRImageFrameDelay(string FileName, int ImageIndex, int FrameIndex)
        {
            int functionReturnValue = 0;
            Decoder objDecoder = new Decoder();
            objDecoder.OpenXPR(FileName);
            functionReturnValue = objDecoder.get_GetImage(ImageIndex).Frames[FrameIndex].Delay;
            objDecoder.CloseXPR();
            return functionReturnValue;
        }

        public Image GetXPRImageFrame(string FileName, int ImageIndex, int FrameIndex)
        {
            Image functionReturnValue = null;
            Decoder objDecoder = new Decoder();
            objDecoder.OpenXPR(FileName);
            functionReturnValue = objDecoder.get_GetImage(ImageIndex).Frames[FrameIndex].Frame;
            objDecoder.CloseXPR();
            return functionReturnValue;
        }

        public void SaveXPRImage(string FileName, int ImageIndex, string DestFileName)
        {
            Decoder objDecoder = new Decoder();
            objDecoder.OpenXPR(FileName);
            objDecoder.SaveImage(ImageIndex, DestFileName);
            objDecoder.CloseXPR();
        }

        public void SaveXPRImageFrame(string FileName, int ImageIndex, int FrameIndex, string DestFileName)
        {
            Decoder objDecoder = new Decoder();
            objDecoder.OpenXPR(FileName);
            objDecoder.get_GetImage(ImageIndex).Frames[FrameIndex].Frame.Save(DestFileName);
            objDecoder.CloseXPR();
        }

        #endregion

        #region "GIF Functions"

        public int GetGIFImageLoops(string FileName) {
        int functionReturnValue = 0;

        System.Drawing.Image objImage = System.Drawing.Image.FromFile(FileName);
        FrameDimension objDimension = new FrameDimension(objImage.FrameDimensionsList[0]);
        
        functionReturnValue = 0;
        
        foreach (PropertyItem objPropertyItem in objImage.PropertyItems) {
            switch (objPropertyItem.Id) {
                case 20737:
                    functionReturnValue = System.BitConverter.ToInt16(objImage.GetPropertyItem(objPropertyItem.Id).Value, 0);
                    break;
            }
        }
        
        objImage.Dispose();
        return functionReturnValue;
        
    }

        public int GetGIFImageFrameCount(string FileName)
        {
            int functionReturnValue = 0;

            System.Drawing.Image objImage = System.Drawing.Image.FromFile(FileName);
            FrameDimension objDimension = new FrameDimension(objImage.FrameDimensionsList[0]);
            functionReturnValue = objImage.GetFrameCount(objDimension);
            objImage.Dispose();
            return functionReturnValue;

        }

        public int GetGIFImageFrameDelay(string FileName, int FrameIndex)
    {
        int functionReturnValue = 0;
        
        byte[] bytDelays = null;
        System.Drawing.Image objImage = System.Drawing.Image.FromFile(FileName);
        FrameDimension objDimension = new FrameDimension(objImage.FrameDimensionsList[0]);
        
        foreach (PropertyItem objPropertyItem in objImage.PropertyItems)
        {
            switch (objPropertyItem.Id) {
                case 20736:
                    bytDelays = objImage.GetPropertyItem(objPropertyItem.Id).Value;
                    break;
            }
        }
        
        if (bytDelays == null) {
            functionReturnValue = 0;
        }
        else {
            functionReturnValue = System.BitConverter.ToInt32(bytDelays, FrameIndex * 4) * 10;
        }
        
        objImage.Dispose();
        return functionReturnValue;
        
    }

        public Image GetGIFImageFrame(string FileName, int FrameIndex)
        {
            Image functionReturnValue = null;

            System.Drawing.Image objImage = System.Drawing.Image.FromFile(FileName);
            FrameDimension objDimension = new FrameDimension(objImage.FrameDimensionsList[0]);
            System.IO.MemoryStream objMemoryStream = new System.IO.MemoryStream();

            objImage.SelectActiveFrame(objDimension, FrameIndex);
            objImage.Save(objMemoryStream, System.Drawing.Imaging.ImageFormat.Png);
            functionReturnValue = System.Drawing.Image.FromStream(objMemoryStream);

            objImage.Dispose();
            return functionReturnValue;

        }

        #endregion

        #region "XML Functions"

        public string XMLClean(string XML)
        {
            //Dim objXMLTools As New XMLTools
            //Return objXMLTools.CleanXML(XML)
            MessageBox.Show("XBMCClean Function not Written");
            return "FAKE";
        }

        public bool XMLValidate(string XML)
        {
            return false;
        }

        public string XMLMergeIncludes(string IncludeXML, string XML)
        {
            return "FAKE";
        }

        public string XMLMergeReferences(string ReferencesXML, string XML)
        {
            return "FAKE";
        }

        #endregion

        #region "Settings Functions"
        
        public void LoadSettings()
        {
           //objSettings.Load(AppPath() + "Settings.xml");
        }

        public void SaveSettings()
        {
            //objSettings.Save(AppPath() + "Settings.xml");
        }

        public string GetSetting(string SettingName)
        {
            //return objSettings.GetSetting(SettingName);
            return "";
        }

        public void SetSetting(string SettingName, string Value)
        {
            //objSettings.SetSetting(SettingName, Value);
        }

        public string GetPluginSetting(string PluginName, string SettingName)
        {
            //return objSettings.GetPluginSetting(PluginName, SettingName);
            return "";
        }

        public void SetPluginSetting(string PluginName, string SettingName, string Value)
        {
            //objSettings.SetPluginSetting(PluginName, SettingName, Value);
        }
        #endregion

    }
}
